src/harness/reference_models/ppa_model/ppa_ref_model.py

```diff
@@ -0,0 +1,82 @@
+#    Copyright 2016 SAS Project Authors. All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+from concurrent.futures import ProcessPoolExecutor as Pool
+import multiprocessing
+from utils import *
+from compute import load_partials
+from reference_models.geo.vincenty import GeodesicPoint
+
+THRESHOLD = -96
+
+
+def __compute_points_gain_prop(device):
+  pool = Pool(multiprocessing.cpu_count())
+  install_param = device['installationParam']
+  # Get all the Function Partials
+  great_circle_partial, antenna_gain_partial, \
+  prop_partial, cnt_partial = load_partials(install_param, THRESHOLD)
+  # Compute all the Points in 0-359 every 200m upto 40km
+  points = pool.map(great_circle_partial, range(0, 360))
+  # Compute the Gain for all Direction
+  antenna_gain = pool.map(antenna_gain_partial, range(0, 360))
+  # Compute the Path Loss
+  path_loss = pool.map(prop_partial, enumerate(points))
+  # Compute the Contour based on Gain and Path Loss Comparing with Threshold
+  cnt = pool.map(cnt_partial, zip(antenna_gain, path_loss))
+  # Smoothing Contour using Hamming Filter
+  cnt = smooth(np.asarray(list(cnt)))
+  print len(cnt)
+  # Generating lat, lon for Contours
+  lat = []
+  lon = []
+  for c, az in zip(cnt, range(0, 360)):
+    l, lo, az = GeodesicPoint(install_param['latitude'],
+                                 install_param['longitude'],
+                                 c, float(az))
+    lat.append(l)
+    lon.append(lo)
+  polygon = geojson.Polygon([zip(lon,lat)])
+  return polygon
+
+
+def ppa_creation_model(device_filenames, pal_record_filenames,
+                       pal_user_id, pal_low_frequency, pal_high_frequency):
+  """Create a PPA Polygon based on the PAL Records and Device Information
+  Args:
+    device_filenames: (List) A list containing device filenames.
```

Comment:
We need to take a list of registrationRequests and palRecords, not the filenames. Loading by filename (especially considering the post-processing that is done later) is not flexible enough to do what is needed.

Reply:
To be clear: we *cannot* be doing the "make PAL records consistent" kind of work in this reference model -- the behavior is totally unexpected and it will not match what the SAS UUT will do when given the same data (which is the definition of a reference model).

Reply:
Instead of doing it in the reference model, do you think it will be a good idea to create a utility function as a part of a reference model, which will call the `makePpaAndPalRecordsConsistent` and `getRandomLatLongInPolygon`? Or leave it up to the Test Dev to call these functions if necessary? If that is the case then the reference model will always assume that the dictionaries provided are always correct(we can have some checks in place).

Reply:
Kate, based on your comment we will need some refactoring and some thinking as to how to return the updated Device Info. Since in the Reference Model itself there’s function makePpaAndPalRecordsConsistent (Janet and Sushant had a long discussion with you regarding the implementation of this function) which basically add all the required parameters for the PPA and PAL. If we don’t do it in reference model, then every time the PPA Reference Model is called we will have to call this function and then we will have to pass the PPA and PAL Record Dictionary as a parameter instead of file names. Separating the two can potentially cause misusing of the functions when someone new to the test code development and want to call the reference model without knowing the fact that the two must be together and the  makePpaAndPalRecordsConsistent function must be call ahead of the reference mode call.

We thought of having it more unified, but if your suggested way is a better solution for good reasons, we will modify the code then.

Thanks

Reply:
James,

I agree that the function makePpaAndPalRecordsConsistent() is a very useful function and that it should be used in the test cases. However, it is inappropriate to use it in a reference model.

A reference model should, by definition, have the same behavior as a "model" SAS when given the same data. If you tried to claim a PPA with conflicting PAL records, a correct SAS should give an error rather than silently aligning the data and ignoring the problem.

It is the responsibility of the code/configuration which calls the reference model to make sure that the inputs make sense. If the inputs do not make sense, the reference model should return an error and exit. The reference model should do *exactly* what a SAS would do in the same situation, full stop. If it does anything else, it is no longer a reference model.

Based on his comment, I think Sushant understands what I'm trying to say (i.e. that it's test dev's responsibility to call these helper functions before calling the reference model, and that the reference model should do at least some error checking) so perhaps he can explain further if you still have questions.

Kate